kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h

```diff
@@ -0,0 +1,197 @@
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Thomas Oberbichler
+//                   Tobias Teschemacher
+//
+
+#if !defined(KRATOS_PROJECTION_NURBS_GEOMETRY_UTILITIES_H_INCLUDED)
+#define KRATOS_PROJECTION_NURBS_GEOMETRY_UTILITIES_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "containers/array_1d.h"
+
+namespace Kratos
+{
+    template<int TDimension, class TPointType> class NurbsCurveGeometry;
+    template<int TDimension, class TPointType> class NurbsSurfaceGeometry;
+    namespace ProjectionNurbsGeometryUtilities
+    {
+        typedef array_1d<double, 3> CoordinatesArrayType;
+
+        /*
+        * @brief Returns the projection of a point onto a Nurbs curve
+        *        geometry using the Newton-Rapshon iterative method
+        * @param rParameter Intial guess for the Newton-Rapshon algorithm
+        *        overwritten by the local coordinates of the projected point onto
+        *        the Nurbs curve geometry
+        * @param rPoint The point to be projected onto the Nurbs curve geometry
+        *        This is overwritten by the Cartesian coordinates of the projected
+        *        point in case the projection is successful 
+        * @param rResult The projection onto the Nurbs curve geometry
+        * @param rNurbsCurve The Nurbs curve geometry onto which the point is 
+        *        to be projected
+        * @param MaxIterations Maximum number of iterations for the Newton-Rapshon 
+        *        algorithm
+        * @param ModelTolerance Tolerance of the CAD model
+        * @param Accuracy Accuracy for the the Newton-Rapshon algorithm
+        */
+        template <int TDimension, class TPointType>
+        bool NewtonRaphsonCurve(
+            CoordinatesArrayType& rParameter,
+            const CoordinatesArrayType& rPoint,
+            CoordinatesArrayType& rResult,
+            const NurbsCurveGeometry<TDimension, TPointType>& rNurbsCurve,
+            const int MaxIterations = 20,
+            const double Accuracy = 1e-6,
+            const double ModelTolerance = 1e-6)
+        {
+            // Intialize variables
+            double residual, delta_t;
+
+            // Loop over all Newton-Raphson iterations
+            for (int i = 0; i < MaxIterations; ++i) 
+            {
+                // Compute the position, the base and the acceleration vector
+                std::vector<array_1d<double, 3>> derivatives;
+                rNurbsCurve.GlobalSpaceDerivatives(
+                    derivatives,
+                    rParameter,
+                    2);
+                rResult = derivatives[0];
+
+                // Compute the distance vector between the point and its 
+                // projection on the curve
+                array_1d<double, 3> distance_vector = rResult - rPoint;
+                if (norm_2(distance_vector) < std::max(ModelTolerance,Accuracy))
```

Comment:
Is this correct to mix the metric ModelTolerance with gradient base Accuracy in a convergence test?

Reply:
Thanks @pooyan-dadvand for the comment. What we do here is basically saying that if the projection distance is smaller than either of the model tolerance or the accuracy then the algorithm is assumed to have converged. Would you recommend something else instead? 

Should we use only the accuracy at this check and in case the algorithm does not converge up to the specified number of iterations then check using the tolerance of the model?

Open for suggestions!!

Reply:
Yes that's a good idea! However I would suggest that this check is then done from outside. So if the solution did not converge one can check if its still acceptable because it's within the model tolerance.